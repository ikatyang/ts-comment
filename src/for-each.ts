import * as ts from 'typescript';

// modified from https://github.com/eslint/typescript-eslint-parser/blob/f5fcc87/lib/convert-comments.js

export type ForEachCallback = ForEachCallbackX<void | boolean>;
export type ForEachCallbackX<T> = (comment: string, scanner: ts.Scanner, source_file: ts.SourceFile) => T;

export const for_each = (source: string | ts.SourceFile, callback: ForEachCallback) => {
  const source_file = (typeof source === 'string')
    ? ts.createSourceFile('', source, ts.ScriptTarget.Latest, false)
    : source;
  const source_text = source_file.getFullText();
  const scanner = ts.createScanner(ts.ScriptTarget.Latest, false, ts.LanguageVariant.Standard, source_text);

  let token = scanner.scan();
  while (token !== ts.SyntaxKind.EndOfFileToken) {
    const start = scanner.getTokenPos();
    const end = scanner.getTextPos();

    switch (token) {
      case ts.SyntaxKind.SingleLineCommentTrivia:
      case ts.SyntaxKind.MultiLineCommentTrivia: {
        const comment = scanner.getTokenText();
        if (callback(comment, scanner, source_file) === false) {
          return;
        }
        break;
      }
      case ts.SyntaxKind.CloseBraceToken: {
        const container = get_node_container(source_file, start, end);
        if (container && (
            container.kind === ts.SyntaxKind.TemplateMiddle ||
            container.kind === ts.SyntaxKind.TemplateTail
          )) {
          token = scanner.reScanTemplateToken();
          continue;
        }
        break;
      }
      case ts.SyntaxKind.SlashToken:
      case ts.SyntaxKind.SlashEqualsToken: {
        const container = get_node_container(source_file, start, end);
        if (container && container.kind === ts.SyntaxKind.RegularExpressionLiteral) {
          token = scanner.reScanSlashToken();
          continue;
        }
        break;
      }
      default:
        // do nothing
        break;
    }
    token = scanner.scan();
  }
};

function get_node_container(source_file: ts.SourceFile, start: number, end: number) {
  let container: undefined | ts.Node;
  find_container(source_file, start, end, target_container => {
    container = target_container;
  });
  return container;
}

function find_container(node: ts.Node, start: number, end: number, callback: (container: ts.Node) => void) {
  const node_start = node.pos;
  const node_end = node.end;
  if (start >= node_start && end <= node_end) {
    if (is_token(node)) {
      callback(node);
    } else {
      ts.forEachChild(node, child_node => find_container(child_node, start, end, callback));
    }
  }
}

function is_token(node: ts.Node) {
  return node.kind >= ts.SyntaxKind.FirstToken && node.kind <= ts.SyntaxKind.LastToken;
}
