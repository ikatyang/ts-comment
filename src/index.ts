export * from './for-each';
export * from './gets';
