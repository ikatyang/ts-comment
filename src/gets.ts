import * as _ts from 'typescript';
import { for_each, ForEachCallbackX } from './for-each';

export type GetsCallback<T> = ForEachCallbackX<T>;

/**
 * get comments from source, optional callback for custom results
 */
export function gets(source: string | _ts.SourceFile): string[];
export function gets<T>(
  source: string | _ts.SourceFile,
  callback: GetsCallback<T>,
  ts?: typeof _ts,
): T[];
export function gets(
  source: string | _ts.SourceFile,
  callback: GetsCallback<any> = comment => comment,
  // istanbul ignore next
  ts = _ts,
): any[] {
  const results: any[] = [];

  for_each(
    source,
    (comment: string, scanner: _ts.Scanner, source_file: _ts.SourceFile) => {
      const result = callback(comment, scanner, source_file);
      results.push(result);
    },
    ts,
  );

  return results;
}
