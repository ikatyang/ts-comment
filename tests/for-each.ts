import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import {for_each} from '../src/index';

const load_fixture = (name: string) =>
  fs.readFileSync(path.resolve(__dirname, `../fixtures/${name}.ts`), 'utf8');

it('should work correctly with source (SourceFile)', () => {
  const source_file = ts.createSourceFile('', load_fixture('general'), ts.ScriptTarget.Latest, false);
  const comments: string[] = [];
  for_each(source_file, comment => {
    comments.push(comment);
  });
  expect(comments).toMatchSnapshot();
});

it('should work correctly with source (string)', () => {
  const comments: string[] = [];
  for_each(load_fixture('general'), comment => {
    comments.push(comment);
  });
  expect(comments).toMatchSnapshot();
});

it('should break loop with callback return false', () => {
  const comments: string[] = [];
  for_each(load_fixture('general'), comment => {
    comments.push(comment);

    // tslint:disable-next-line:no-magic-numbers
    return (comments.length < 5);
  });

  expect(comments).toMatchSnapshot();
});

it('should parse correctly with Template Literal', () => {
  const comments: string[] = [];
  for_each(load_fixture('template'), comment => {
    comments.push(comment);
  });
  expect(comments).toMatchSnapshot();
});

it('should parse correctly with Regex Literal', () => {
  const comments: string[] = [];
  for_each(load_fixture('regex'), comment => {
    comments.push(comment);
  });
  expect(comments).toMatchSnapshot();
});
