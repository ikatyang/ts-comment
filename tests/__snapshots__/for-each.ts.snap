// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should break loop with callback return false 1`] = `
Array [
  "// single-line-comment 1",
  "/* multi-line-comment 1 */",
  "// single-line-comment 2",
  "/* multi-line-comment 2 */",
  "// single-line-comment 3",
]
`;

exports[`should parse correctly with Regex Literal 1`] = `
Array [
  "// single-line-comment 1",
  "/* multi-line-comment 1 */",
  "// single-line-comment 2",
  "/* multi-line-comment 2 */",
  "// single-line-comment 3",
  "/* multi-line-comment 3 */",
  "// single-line-comment 4",
  "/* multi-line-comment 4 */",
  "// single-line-comment 5",
  "/* multi-line-comment 5 */",
]
`;

exports[`should parse correctly with Template Literal 1`] = `
Array [
  "// single-line-comment 1",
  "/* multi-line-comment 1 */",
  "// single-line-comment 2",
  "/* multi-line-comment 2 */",
  "// single-line-comment 3",
  "/* multi-line-comment 3 */",
  "// single-line-comment 4",
  "/* multi-line-comment 4 */",
  "// single-line-comment 5",
  "/* multi-line-comment 5 */",
]
`;

exports[`should work correctly with source (SourceFile) 1`] = `
Array [
  "// single-line-comment 1",
  "/* multi-line-comment 1 */",
  "// single-line-comment 2",
  "/* multi-line-comment 2 */",
  "// single-line-comment 3",
  "/* multi-line-comment 3 */",
  "// single-line-comment 4",
  "/* multi-line-comment 4 */",
  "// single-line-comment 5",
  "/* multi-line-comment 5 */",
]
`;

exports[`should work correctly with source (string) 1`] = `
Array [
  "// single-line-comment 1",
  "/* multi-line-comment 1 */",
  "// single-line-comment 2",
  "/* multi-line-comment 2 */",
  "// single-line-comment 3",
  "/* multi-line-comment 3 */",
  "// single-line-comment 4",
  "/* multi-line-comment 4 */",
  "// single-line-comment 5",
  "/* multi-line-comment 5 */",
]
`;
